### PR TITLE
Calculate CI with distributions when scrolled into view

### PR DESCRIPTION
**Pull request by musicin3d**
At the moment, this is just a proof of concept. It loads distributions, calculates the CI, and dumps the information to the page. It also highlights the product tile when it's finished.

There are a few bugs that still need to be worked out.

### CODE REVIEW

**Comment by chrismbryant** [comment]
This is pretty cool! I like the functionality

**Comment by musicin3d** [comment]
I reorganized inject.js, so it with conflict with Chris's pull request... https://github.com/chrismbryant/amazon-confidence-interval/pull/19

**Comment by chrismbryant** [comment]
> I reorganized inject.js, so it with conflict with Chris's pull request... #19

The changes I made to `inject.js` were pretty minor, so the conflicts should be pretty easy to resolve

**Comment by musicin3d** [comment]
Resolved conflicts and fixed bugs. This is ready for final review.

Tagging: https://github.com/chrismbryant/amazon-confidence-interval/issues/9

**Comment by musicin3d** [comment]
I don't know if this is a regression or a new bug. At https://www.amazon.com/s?me=A3EPIN8KW1GGQX&marketplaceID=ATVPDKIKX0DER the scraping fails almost entirely.

Edit: It's a regression. I'll fix it.

**Comment by aeciorc** [comment]
Sorry about delay in reviewing this, should be able to check it out this evening

File: src/shared/scraper.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    async findAllProducts(){
+        return new Promise((resolve, reject) => {
+            if (document.readyState == "loading"){
+                document.addEventListener('DOMContentLoaded', run)
+            } else {
+                run()
+            }
+            function run(){
+                //TODO sometimes products are featured together, and this selects the entire block instead of each product
+                const images = document.querySelectorAll('[data-component-type=s-product-image]')
```
Comment (aeciorc) [inline src/shared/scraper.js:26]: Nice, I didn't know you could select based on attributes

File: src/shared/scraper.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    async findAllProducts(){
+        return new Promise((resolve, reject) => {
+            if (document.readyState == "loading"){
+                document.addEventListener('DOMContentLoaded', run)
+            } else {
+                run()
+            }
+            function run(){
+                //TODO sometimes products are featured together, and this selects the entire block instead of each product
```
Comment (aeciorc) [inline src/shared/scraper.js:25]: I noticed it also doesn't work when you're searching within a category, eg: https://www.amazon.ca/s?k=notes&i=alexa-skills&ref=nb_sb_noss

File: src/chrome/inject.js [hunk 3]
```diff
@@ -1,46 +1,113 @@
...
-            await new Promise(r => setTimeout(r, 500));
-            productsDOM = document.querySelectorAll(productsSelector);
+function handleScroll(evt){
+    if(scrollTicking)  return
+    scrollTicking = true
+    window.requestAnimationFrame(()=>{
+        const leftover = []
+        for(let product of noDistProducts) {
+            if (productInView(product)){
+                // noinspection JSIgnoredPromiseFromCall
```
Comment (aeciorc) [inline src/chrome/inject.js:73]: shouldn't you also check for `product.reviews`? I noticed that in products without reviews, the avg viz doesn't get added, but a blank dist viz does

File: src/shared/scraper.js (hunk 2 above)
Comment (aeciorc) [reply src/shared/scraper.js:25]: Looks like this is the same issue you pointed out earlier

File: src/shared/scraper.js (hunk 1 above)
Comment (musicin3d) [reply src/shared/scraper.js:26]: :grin:

File: src/chrome/inject.js (hunk 3 above)
Comment (musicin3d) [reply src/chrome/inject.js:73]: Good find! I've pushed an update that filters out those products before setting `noDistProducts`

**Comment by musicin3d** [comment]
I'll try to resolve all issues tomorrow night.

**Comment by musicin3d** [comment]
Closing due to the revelation in https://github.com/chrismbryant/amazon-confidence-interval/issues/10#issuecomment-611661703

**Comment by chrismbryant** [comment]
@musicin3d you don’t think it still might be worth taking this approach even if we do use the average star rating? Especially if we want to offer a mode where the visualization is injected into the popover?

**Comment by musicin3d** [comment]
Inserting things into the DOM isn't very expensive. In fact, it's probably
be better to do it all at once whenever possible. If the calculations are
cheap as well, then I see no point in using the scroll event. My motivation
behind going down this path was that retrieving the individual ratings
required network requests. Those are expensive, and they were even being
throttled in this case.

If we're expecting some expensive calculations, then there would still be
some value in this. If we are not then I believe we can find a better way.
For example, we can simply insert a score under each rating, and we can
display our own popover when you mouse over that.

If we don't like that then we might be able to find a way to modify
Amazon's popover when it's created, but that could get messy regardless of
using the scroll event or not.

To get back out of the weeds... Are we expecting expensive calculations?

**Comment by chrismbryant** [comment]
I see, thanks for clarifying. Yeah I don’t imagine we’d be doing any expensive calculations

**Comment by aeciorc** [comment]
Hey @musicin3d, the changes you made in refactoring the scraping are still interesting even if didn't end up benefitting from using the distributions. So I think there's value in merging some of the commits in this PR to master 😛

**Comment by musicin3d** [comment]
Hey! I appreciate that. 😄 I'll go back through and make a new pull request
with the refactoring.

**Pull request closed without merging**